pkg/tools/shodan.go

```diff
@@ -0,0 +1,150 @@
+					"asn":      map[string]any{"type": "string"},
+					"hostname": map[string]any{"type": "string"},
+					"org":      map[string]any{"type": "string"},
+					"net":      map[string]any{"type": "string"},
+					"country":  map[string]any{"type": "string"},
+					// Port list – a slice of ints
+					"port": map[string]any{"type": "array",
+						"items": map[string]any{"type": "integer"},
+					},
+				},
+			},
+
+			// Top‑level flags (page, filters…)
+			"params": map[string]any{
+				"type": "object",
+				"properties": map[string]any{
+					"page": map[string]any{
+						"type":        "uint",
+						"description": "Page number of the search (default 1).",
+					},
+				},
+			},
+		},
+	},
+}
+
+// -------------------------------------------------------------------
+// 2️⃣  Args struct – matches the JSON we described above.
+// -------------------------------------------------------------------
+type ShodanToolArgs struct {
+	Query  search.Query  `json:"query"`
+	Params search.Params `json:"params,omitempty"`
+}
+
+// -------------------------------------------------------------------
+// 3️⃣  Tool type (empty; only used for method receiver).
+// -------------------------------------------------------------------
+type ShodanTool struct{}
+
+// -------------------------------------------------------------------
+// 4️⃣  Call – receives the JSON payload, runs the shodan client and returns a string.
+// -------------------------------------------------------------------
+func (s ShodanTool) Call(ctx context.Context, input string) (string, error) {
+	// 1. Unmarshal incoming JSON into our args struct
+	shodanArgs := ShodanToolArgs{}
+	if err := json.Unmarshal([]byte(input), &shodanArgs); err != nil {
+		return "", fmt.Errorf("json unmarshal: %w", err)
+	}
+
+	// 2. Build a client – the library expects an API key in an env var.
+	client, _ := shodanclient.GetClient(
+		os.Getenv("SHODAN_API_KEY"), // <-- your key
+		http.DefaultClient,
+		true, // use API v1
+	)
+	//ctx = context.Background()
+
+	// 3. Build a search request from the supplied args.
+	shodanSearch := search.Params{
+		Page:  shodanArgs.Params.Page,
+		Query: shodanArgs.Query,
+	}
+
+	result, err := client.Search(ctx, shodanSearch)
+	if err != nil {
+		return "", fmt.Errorf("shodan search failed: %w", err)
+	}
+
+	// 4. Return the raw JSON string so that other tools can consume it.
+	out, err := json.Marshal(result)
+	if err != nil {
+		return "", fmt.Errorf("json marshal: %w", err)
+	}
+	return string(out), nil
+}
+
+// -------------------------------------------------------------------
+// 5️⃣  Register this tool in the global registry (used by NewToolsExecutor).
+// -------------------------------------------------------------------
+func init() {
+	globalToolsRegistry = append(globalToolsRegistry,
+		func(ctx context.Context, cfg config.Config) (*tools.ToolData, error) {
+			// Skip if you disable it via an env var
+			if cfg.ShodanDisable {
+				return nil, nil
+			}
+
```

Comment:
cfg already have Shodan_key field (do not use underscores in go), and it should be passed to ShodanTool{} here, with error return if it's empty (see semantic search for example)